### ISSUE TITLE
Recommended ActiveJob integration

### ISSUE THREAD

**Issue by monkbroc** [rollbar/rollbar-gem#263]
Since ActiveJob has been part of Rails for a while now it would be great if the README included a recommended integration between Rollbar and ActiveJob so that any uncaught exception in a background job is reported.

I'm using SuckerPunch as the backend and I thought that `config.use_sucker_punch` would report the errors, but it only queues a job through sucker punch to report already caught errors.

Here's my integration. Any better ideas?

```ruby
class BaseJob < ActiveJob::Base
  # Report any errors in jobs to Rollbar
  rescue_from Exception do |exception|
    Rollbar.error(exception, job: self.class.name, job_id: job_id)
  end
end
```

**Comment by jondeandres** [comment]
@monkbroc that snippet should be enough, yes. Maybe we should comment it in our README.

We could provide a module that the users will include in their jobs, but I'm not sure if this will be more complex for them or it's better just let them send the errors manually.

@monkbroc do you have any preference about this?

Would you like this?

```ruby
class YourAwesomeJob < ActiveJob::Base
  include Rollbar::ActiveJob
end
```

**Comment by monkbroc** [comment]
A module with a working default would be great.

I made a PR #264 for integrating with ActiveJob.

**Comment by DMA57361** [comment]
I've just tried this out - to replace some custom error handling we had that did a similar job - and thought I'd point out that this can cause some unexpected behaviour because it swallows unhandled exceptions.

I noticed this because we use Sidekiq as our ActiveJob backend and it will automatically retry failed jobs. However, this stopped happening when I activated this because the exceptions were caught and the job was instead considered to have been successful.

Since we've only the one project using ActiveJob at the moment I don't know what effect, if any, this could have on most of the other backends. It also seems to cause jobs called with `perform_now` or via the inline backend to fail for unhandled exceptions (which they would otherwise - the exception is just allowed to bubble up),

I think either a reraise or maybe warning in the readme would be useful.

**Comment by jondeandres** [comment]
Hi @DMA57361, thanks for the tip.

I think this kind of behaviors will be a problem between the ActiveJob API and the backends. I'm not sure about which solution is the best. I'll investigate a bit about this.

The philosophical problem I find is if, in your case, Sidekiq should determine if it should retry or not, when ActiveJob has its own API to define that.

And I'm not sure about the reraise inside `rescue_from`. I'll test it. Maybe that's the best solution yes.

**Comment by DMA57361** [comment]
Yes, us catching the exceptions then using `retry_job` in some way would make our retrying requirements more explicit and more backend agnostic.

To be honest, I hadn't realised `retry_job` was a thing until I looked for it just now; my experience of ActiveJob is limited to the only last few weeks - previously we've just used Resque or Sidekiq directly. That said, I think I'll stick with the Sidekiq retrying for now, it seems more featureful, but that's my personal choice so don't worry about it.

Really, I just thought I'd make sure the potential problem was known, so I'll leave this with you to decide what's best.

**Comment by jondeandres** [comment]
Yah...this sounds like a conflict between backend features and ActiveJob. I'm thinking what to do in this case. We'll contact you when we decide something ok?

Maybe for you the best is use our Sidekiq integration instead of ActiveJob.

**Comment by DMA57361** [comment]
Oh, didn't realise you had an integration for catching errors within Sidekiq - I assume you mean this https://github.com/rollbar/rollbar-gem/blob/master/lib/rollbar/sidekiq.rb? I'd not seen that before, is it documented somewhere?

I will take a look. Previously we've just added a Rollbar.error call to sidekiq's error_handlers, which is pretty much the same.

**Comment by jondeandres** [comment]
Yes @DMA57361, you have some docs here https://github.com/rollbar/rollbar-gem#using-sidekiq.

If you have any doubt write us :).

Cheers

**Comment by DMA57361** [comment]
I think that's for using Sidekiq to make reports to Rollbar asynchronously, rather than using Rollbar to automatically capture Sidekiq failures.

**Comment by jondeandres** [comment]
Oh yes, sorry, I mixed the things.

You just need to `require 'rollbar/sidekiq'`, yes. It's basically a `Rollbar.error` but sending some context data.

**Comment by monkbroc** [comment]
Good point about swallowing the exception. I agree that reraising would
make more sense.

**Comment by brianr** [comment]
#270 has been merged and released (version 2.1.1). Thanks @monkbroc and @DMA57361!
